implement Window function definitions revised

greenTara opened the pull request:
updates to the definitions of window functions initiatied as discussed in https://github.com/streamreasoning/RSP-QL/issues/33

greenTara committed:
implement Window function definitions revised

updates to the definitions of window functions initiatied as discussed inhttps://github.com/streamreasoning/RSP-QL/issues/33

greenTara commented:
Minh - would you be willing to review this pull request?

There are still some things about the count-based window functions that I'm not happy with, as mentioned in the ednotes.

daotranminh commented:
Hi Tara, I will have a look.

Thanks,
Minh.

On Thu, Feb 11, 2016 at 3:11 AM, Tara Athan notifications@github.com
wrote:

> Minh - would you be willing to review this pull request?
> 
> There are still some things about the count-based window functions that
> I'm not happy with, as mentioned in the ednotes.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/streamreasoning/RSP-QL/pull/42#issuecomment-182673122
> .

## 

DAO Tran Minh
http://web.student.tuwien.ac.at/%7Ee0727429

greenTara commented:
This pull request is also related to #36

greenTara committed:
replace "convex substream" with "window"

also add qualifier "temporal" to count-based

fixed typo "symmetric" => "antisymmetric"

greenTara committed:
define the term window relation

greenTara committed:
add a definition for length-based window relation

greenTara committed:
clarify the definition of "cover"

greenTara committed:
change "p" to "P" because HTML does a bad job of typesetting code with letters that drop below the baseline.

greenTara committed:
Merge branch 'gh-pages' into issue33

* gh-pages:
  improve non-literal example with real predicate
  add non-literal example
  make the clause about sequences of graphs without timestamps normative.
  expand explanation of example
  modify terminology "explicit" => "literal-" and add clarification
  add ednotes about context
  add quote of RDF Dataset semantics section
  Edited query operator section: stated as status of desired RSP-QL operators
  formatted and removed the definitions of RSP model. RSP model reshaped
  implementation of https://github.com/streamreasoning/RSP-QL/issues/28

greenTara committed:
Merge branch 'gh-pages' into issue33

* gh-pages:
  correcting a typo
  cross-referencing to window-functions
  implementing https://github.com/streamreasoning/RSP-QL/issues/31

greenTara committed:
Merge pull request #4 from streamreasoning/25.2

Addressed latest feedback

greenTara committed:
Merge branch 'gh-pages' of https://github.com/greenTara/RSP-QL into 25.2

greenTara committed:
Merge branch 'issue33' of https://github.com/greenTara/RSP-QL into issue33

greenTara committed:
Merge pull request #5 from streamreasoning/gh-pages

Merge

greenTara committed:
Merge branch 'gh-pages' of https://github.com/greenTara/RSP-QL into 25.2

greenTara committed:
Merge branch 'gh-pages' of https://github.com/greenTara/RSP-QL into issue33

greenTara committed:
Merge pull request #6 from streamreasoning/25.2

Merge branch 'gh-pages' of https://github.com/streamreasoning/RSP-QL …

greenTara committed:
Merge branch '25.2' of https://github.com/greenTara/RSP-QL into issue33

greenTara committed:
Merge pull request #7 from streamreasoning/gh-pages

Pull

greenTara committed:
Merge branch 'gh-pages' of https://github.com/greenTara/RSP-QL into issue33

dellaglio commented on Abstract Syntax and Semantics Document/index.html:
```diff
@@ -500,53 +500,159 @@
       
-      <p class="note">In the following, we take the case where temporal entities in the stream are time instants.</p>
+      <p class="ednote">The definition of cover, and the following discussion of it, could probably be moved to follow shortly after the definition of timestamped graph so that
+      this concept could be used in other definition.</p>
```
+1

dellaglio commented on Abstract Syntax and Semantics Document/index.html:
```diff
@@ -500,53 +500,159 @@
+      </p>
+      <p class="note">Informally, a window is a temporally-contiguous selection from the original stream, without gaps.</p>
+      
+      <p class="ednote">The definition of window should be moved to the substream section.</p>
```
Probably this part should replace the current substream section...

greenTara replied:
As the ednote says, the plan is to move the definition of window into the substream section, following the definition fo substream.

dellaglio commented on Abstract Syntax and Semantics Document/index.html:
```diff
@@ -500,53 +500,159 @@
+        the timestamp predicate  <code>P</code> of <code>S(i)</code> is in <code>P</code> and
+        <ul>          
+          <li><code>t</code> is the timestamp of <code>S(i)</code>,</li>
+          <li><code>T</code> is the set of temporal entities that are timestamps for elements of <code>S</code>
```
T is used at the beginning to denote the set of temporal entities. Instead of using <code>T</code> here, I would suggest to use a different symbol.

greenTara replied:
Since the set of timestamps in a stream S is a dependent function of the stream S, we could us T(S) instead of T here. Do you think that would be clear enough?

We could also introduce a more special symbol for the set (should actually be class, I think) of all temporal entities, instead of "T". E.g. T<sub>all</sub>

beta2k replied:
+1 for T(S)

dellaglio commented on Abstract Syntax and Semantics Document/index.html:
```diff
@@ -500,53 +500,159 @@
+      (a problem that might be resolved by defining a window relation instead of a function.)
+      If elements are discarded in a nonrandom fashion based on their order in the stream sequence, then this
+      would be a function, but would not preserve isomorphism, and so would not be a window function. 
+      </p>
```
Temporal-count-based windows overcome the non-determinism issue of (what DSMS literature names) count-based windows, and this note clearly explains it. 
This paragraph warns about the non-determinism that may arise in non-temporal count-based window, but I am wondering if we should not add a conclusion. I mean, at the moment we are telling readers that it is up them to decide how to do that, but to be aware that you will have those issues [...]. 
Is this the only advise we are want to give to ppl interested in having windows with an exact number of elements?

greenTara replied:
This note explains that in the most general case, where the temporal entities are not totally ordered or duplicate time stamps are allowed, it is not possible to overcome the non-determinism issue to get a window function that returns an exact number of elements. We can add a forward reference to the profiles, as advice to ppl interested in having windows with an exact number of elements.

You mention that there is some nomenclature from the literature that we are perhaps not consistent with. Should we discuss this in the call?

dellaglio replied:
I like this solution of relating this part to the profile one to explain how to tackle the nondeterminism issue. 

I also like the name, as the "temporal" sorts out the doubts about the meaning of this window w.r.t. the (non-temporal) count-based windows defined in literature. 
Anyway, I think it is worth to give a name to the window with an exact number of elements, as they are relevant and needed in several use cases. What about element-count-based windows?

beta2k replied:
I like element-count-based windows. Would this require a precise definition of what an `element` actually is somehwere?

greenTara replied:
What's in there now is called "length-based window relation" or
"length-based window function". I am happy to change that to
"element-count-based" as it seems more symmetrical to
"temporal-count-based". There is a separate issue, on the other document
actually
https://github.com/streamreasoning/RSP-QL/issues/64#issuecomment-222165487,
about "element" being the terminology for a single "thingie" of the stream
sequence. I suggested "item", "element", or "member". Whatever gets
selected, then we should use that for this window function, I believe.

Tara

On Fri, May 27, 2016 at 3:37 PM, beta2k notifications@github.com wrote:

> In Abstract Syntax and Semantics Document/index.html
> https://github.com/streamreasoning/RSP-QL/pull/42#discussion_r64905218:
> 
> > +
> > -      <p class="note">The substream resulting from the application of a temporal-count-based window function is temporal-count-bounded.</p>
> >   +
> > -      <p class="note">Due to the potential for stream elements with incomparable or duplicate timestamps, the number of elements in the substream having a particular
> > -        timestamp predicate
> > -      is not guaranteed to be equal to the depth specified for the predicate by the temporal-count-based window. </p>
> >   +
> > -      <p class="ednote">
> > -       Applications that require a substream with an exact number of elements for a specified timestamp predicate
> > -       might randomly discard elements after applying a temporal-count-based window function, according to some criterion, e.g.
> > -       extreme elements (maximal or minimal, depending on the orientation of the counting).
> > -      However, this extra step causes the process to be nondeterministic, and hence is not a function
> > -      (a problem that might be resolved by defining a window relation instead of a function.)
> > -      If elements are discarded in a nonrandom fashion based on their order in the stream sequence, then this
> > -      would be a function, but would not preserve isomorphism, and so would not be a window function.
> > -      </p>
> 
> I like element-count-based windows. Would this require a precise
> definition of what an element actually is somehwere?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/streamreasoning/RSP-QL/pull/42/files/06507b6e1f5cec50c3fde474a72885912c68201a#r64905218,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABMMVd5PoGEBYbi5gGBldgCaqK0p2t7aks5qFvOjgaJpZM4HX2bR
> .

dellaglio commented on Abstract Syntax and Semantics Document/index.html:
```diff
@@ -500,53 +500,159 @@
-      <li><code>t'</code> satisfies that <code>#S|<sub>p</sub>[t',t] ≥ l</code> and <code>#S|<sub>p</sub>[t'+1,t] &lt; l</code> </li>
-    </ul>
+      <p class="note">The substream resulting from the application of a time-based window function is time-bounded.</p>
     </section>
```
Are we planning to add examples later?

greenTara replied:
Yes, we will add (at least) two running examples, one leading into the time-series profile and one leading to the linked-list profile. I am waiting until we get these basic definitions stabilized and do some reorganization because otherwise the branches get tangled up.

dellaglio commented:
It looks almost fine to me, I have just some doubts that we could briefly discuss in the next phone call.

kiat commented:
I have a general comment about the terms "substream" and "window". I am not sure if you can see a window as specialization of a substream. 

I document it is defined:

"Substream:
A substream S' of an RDF stream S is an RDF stream that is isomorphic to a subsequence of S.

NOTE:
There are several specializations of the substream relation which are useful in practice. These include the filter relation, where stream elements are selected based on satisfaction of some criterion, and the window relation, where a temporally contiguous portion of the stream is selected."

But a stream is an unbounded sequence of data and a window is a bounded or a finite portion of the stream. Like the following definition given by Chakravarthy et al. 

"Informally, a data stream is defined as an unbounded (or never-ending) sequence of data items that are usually ordered either explicitly by a timestamp 
or by the values of one or more data elements." 
"The notion of a window has been introduced to overcome the blocking aspect of a number of operators. Informally, a window defines a finite portion of the stream (as a relation) for processing purposes." 
"A window is defined as a historical snapshot of a finite portion of a stream at any time point. A window defines the meaningful set of data used by operators to compute their functions." [1]
- If a substream is a stream then it is unbouned. 
- If a window is a substream then it is unbouned but a window is mostly defined as a bouned portion of the stream so that it can not be a stream. 

In my mind a substream is a stream using a filter relation and a window is not a substream. A window will be updated in different iteration based on the window semantics but it does not produce an unbounded stream. 

[1] Stream Data Processing: A Quality of Service Perspective Modeling, Scheduling, Load Shedding, and Complex Event Processing. Chakravarthy, Sharma, Qingchun, Jiang

greenTara commented:
"But a stream is an unbounded sequence of data and a window is a bounded or
a finite portion of the stream"

I don't have access to this book, so I can't see what they provide as the
definition of stream and substream. In any case, we are not following the
definitions of Chakravarthy, so our definitions have different
consequences, and this intuition you have about the terms "substream" and
"window" does not apply, in general, to RDF substreams and windows. In
particular, RDF streams, and their substreams and windows, may be bounded
or unbounded. I believe these aspects of the definitions have already been
approved by the group.

I don't believe that it would be good to restrict substreams to be the
result of some filtering. That is certainly one way to generate substreams,
but it is not the only use for substreams. This would also be unnecessarily
complicated, as we would have to provide a definition of filter.

Conversely, in the very general case we consider, having a requirement that
windows be finite would cause difficulty. As mentioned in the Zeno's
paradox example, if there is not a minimum granularity to the temporal
entities, then even a time-based window with both upper and lower bounds
can not be guaranteed to produce a finite window.

The many explanatory notes that are included in this pull request are
intended to address these questions.

Also, such requirements, e.g. to have windows be finite, are what motivates
the profiles. Once the class of RDF streams being considered is restricted
by additional properties, then it is possible to guarantee that certain
window functions produce finite windows.

Tara

greenTara committed:
add "general" modifier for window functions on all streams

greenTara committed:
add forward link to Profiles section

greenTara commented:
OK, I have made the changes that were discussed. If the new material looks ok, then the pull request can be merged.

dellaglio commented:
Personally, I prefer item, but all of your proposals look fine to me.

Il Ven 27 Mag 2016 18:04 Tara Athan notifications@github.com ha scritto:

> OK, I have made the changes that were discussed. If the new material looks
> ok, then the pull request can be merged.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/streamreasoning/RSP-QL/pull/42#issuecomment-222199661,
> or mute the thread
> https://github.com/notifications/unsubscribe/AD_Toh8y-R6sFs5VITDZCU-JEQtsGmoJks5qFyNcgaJpZM4HX2bR
> .
> 
> ## 

## 

Sent from my Android phone. Please excuse my brevity.

kiat commented:
My point was only that a window is a bounded segment of a stream. 

Data stream processing and CEP community consider a window to be a bounded sequence of data items because of computation purposes.

Like the Window definition given by  "Event Processing Technical Society"

http://www.complexevents.com/wp-content/uploads/2011/08/EPTS_Event_Processing_Glossary_v2.pdf

A stream may or may not be bounded but a window is bounded. 

If you define a window to be a specialization of a substream then it is a bounded substream (stream).

greenTara commented:
A popular form of a window function is to be defined by a temporal
interval, with a lower and upper bound. When this is applied to a general
RDF stream, it does not necessarily produce a bounded substream. Example is
the Zeno's paradox stream.

Even in the case when we don't have such "pathological" behaviour in the
timestamps, a window function can have a lower bound only, no upper bound.

On Fri, May 27, 2016 at 9:20 PM, Kia Teymourian notifications@github.com
wrote:

> My point was only that a window is a bounded segment of a stream.
> 
> Data stream processing and CEP community consider a window to be a bounded
> sequence of data items because of computation purposes.
> 
> Like the Window definition given by "Event Processing Technical Society"
> 
> http://www.complexevents.com/wp-content/uploads/2011/08/EPTS_Event_Processing_Glossary_v2.pdf
> 
> A stream may or may not be bounded but a window is bounded.
> 
> If you define a window to be a specialization of a substream then it is a
> bounded substream (stream).
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/streamreasoning/RSP-QL/pull/42#issuecomment-222232043,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABMMVU_jy11PQws0FAGaXaIHeBaipqfoks5qF0PvgaJpZM4HX2bR
> .
